File: examples/python/m2-tool-runtime/workspace.py
"""工作区路径约束与同目录原子替换：教学级边界，不是 OS 沙箱。"""

from __future__ import annotations

import os
from pathlib import Path
import tempfile


class WorkspaceError(ValueError):
    """路径越界或工作区配置错误。"""


class Workspace:
    """把用户提供的相对/绝对路径解析并限制在 root 内。"""

    def __init__(self, root: str | Path) -> None:
        self.root = Path(root).resolve()
        if not self.root.exists():
            raise WorkspaceError(f"workspace root does not exist: {self.root}")
        if not self.root.is_dir():
            raise WorkspaceError(f"workspace root is not a directory: {self.root}")

    def resolve(self, user_path: str | None = None, *, must_exist: bool = False) -> Path:
        """解析路径，并检查现有路径或最近现有父目录的真实位置。

        明确拒绝任何 ``..`` 分量。对于尚不存在的写入目标，从最近的现有
        父目录开始 canonicalize，因此外部 symlink 父目录也不能绕过边界。
        """

        if user_path is None or user_path.strip() == "" or user_path.strip() == ".":
            candidate = self.root
        else:
            raw = Path(user_path)
            if ".." in raw.parts:
                raise WorkspaceError(f"path escapes workspace root: {user_path!r}")
            candidate = raw if raw.is_absolute() else self.root / raw

        if candidate.exists():
            resolved = candidate.resolve(strict=True)
        else:
            missing: list[str] = []
            ancestor = candidate
            while not ancestor.exists():
                if ancestor.parent == ancestor:
                    raise WorkspaceError(f"cannot resolve path: {user_path!r}")
                missing.append(ancestor.name)
                ancestor = ancestor.parent
            resolved = ancestor.resolve(strict=True).joinpath(*reversed(missing))

        try:
            resolved.relative_to(self.root)
        except ValueError as error:
            raise WorkspaceError(
                f"path escapes workspace root: {user_path!r}"
            ) from error

        if must_exist and not resolved.exists():
            raise WorkspaceError(f"path does not exist: {user_path!r}")
        return resolved

    def relative_to_root(self, path: Path) -> str:
        """把绝对路径格式化为相对 root 的 posix 路径。"""

        return path.resolve().relative_to(self.root).as_posix() or "."

    def atomic_write_text(self, path: Path, content: str) -> None:
        """在目标同目录写临时文件，再用 ``os.replace`` 一次替换目标。

        同目录保证替换不会跨文件系统；失败时清理临时文件，原文件保持不变。
        这仍不解决 TOCTOU、挂载点变化或进程权限过大的生产级问题。
        """

        parent = self.resolve(str(path.parent))
        parent.mkdir(parents=True, exist_ok=True)
        checked = self.resolve(str(path))
        temp_name: str | None = None
        try:
            with tempfile.NamedTemporaryFile(
                mode="w",
                encoding="utf-8",
                dir=parent,
                prefix=f".{path.name}.",
                suffix=".tmp",
                delete=False,
            ) as handle:
                temp_name = handle.name
                handle.write(content)
                handle.flush()
                os.fsync(handle.fileno())
            os.replace(temp_name, checked)
            temp_name = None
        finally:
            if temp_name is not None:
                Path(temp_name).unlink(missing_ok=True)

File: examples/rust/m2-tool-runtime/src/workspace.rs
//! 工作区路径约束与同目录原子替换：教学级边界，不是 OS 沙箱。

use crate::tool_types::ToolError;
use std::fs::OpenOptions;
use std::io::Write;
use std::path::{Component, Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};

static TEMP_COUNTER: AtomicU64 = AtomicU64::new(0);

/// 固定 root 的教学级路径边界。
#[derive(Clone, Debug)]
pub struct Workspace {
    root: PathBuf,
}

impl Workspace {
    /// 固定并 canonicalize 工作区根目录。
    pub fn new(root: impl AsRef<Path>) -> Result<Self, ToolError> {
        let root = root
            .as_ref()
            .canonicalize()
            .map_err(|error| ToolError::new(format!("workspace root invalid: {error}")))?;
        if !root.is_dir() {
            return Err(ToolError::new(format!(
                "workspace root is not a directory: {}",
                root.display()
            )));
        }
        Ok(Self { root })
    }

    /// 返回已经 canonicalize 的根目录。
    pub fn root(&self) -> &Path {
        &self.root
    }

    /// 解析路径；现有目标查真实路径，不存在目标查最近的现有父目录。
    pub fn resolve(&self, user_path: Option<&str>, must_exist: bool) -> Result<PathBuf, ToolError> {
        let raw = user_path.unwrap_or(".");
        let path = Path::new(if raw.trim().is_empty() { "." } else { raw });
        if path.components().any(|part| part == Component::ParentDir) {
            return Err(ToolError::new(format!(
                "path escapes workspace root: {user_path:?}"
            )));
        }
        let candidate = if path.is_absolute() {
            path.to_path_buf()
        } else {
            self.root.join(path)
        };
        let resolved = canonicalize_with_missing_tail(&candidate)?;
        if !resolved.starts_with(&self.root) {
            return Err(ToolError::new(format!(
                "path escapes workspace root: {user_path:?}"
            )));
        }
        if must_exist && !resolved.exists() {
            return Err(ToolError::new(format!(
                "path does not exist: {user_path:?}"
            )));
        }
        Ok(resolved)
    }

    /// 以 POSIX 风格返回工作区相对路径。
    pub fn relative_to_root(&self, path: &Path) -> Result<String, ToolError> {
        let abs = canonicalize_with_missing_tail(path)?;
        let rel = abs
            .strip_prefix(&self.root)
            .map_err(|_| ToolError::new("path escapes workspace root"))?;
        let text = rel.to_string_lossy().replace('\\', "/");
        Ok(if text.is_empty() { ".".into() } else { text })
    }

    /// 在目标同目录完整写临时文件，再 rename 替换目标。
    pub fn atomic_write(&self, path: &Path, bytes: &[u8]) -> Result<(), ToolError> {
        let parent = path
            .parent()
            .ok_or_else(|| ToolError::new("target has no parent directory"))?;
        let checked_parent = self.resolve(Some(&parent.to_string_lossy()), false)?;
        std::fs::create_dir_all(&checked_parent)
            .map_err(|error| ToolError::new(error.to_string()))?;
        let checked_path = self.resolve(Some(&path.to_string_lossy()), false)?;
        let file_name = checked_path
            .file_name()
            .and_then(|name| name.to_str())
            .ok_or_else(|| ToolError::new("target file name is not valid UTF-8"))?;
        let nonce = TEMP_COUNTER.fetch_add(1, Ordering::Relaxed);
        let temp =
            checked_parent.join(format!(".{file_name}.{}.{}.tmp", std::process::id(), nonce));

        let result = (|| {
            let mut handle = OpenOptions::new()
                .write(true)
                .create_new(true)
                .open(&temp)
                .map_err(|error| ToolError::new(error.to_string()))?;
            handle
                .write_all(bytes)
                .and_then(|_| handle.sync_all())
                .map_err(|error| ToolError::new(error.to_string()))?;
            std::fs::rename(&temp, &checked_path).map_err(|error| ToolError::new(error.to_string()))
        })();
        if result.is_err() {
            let _ = std::fs::remove_file(&temp);
        }
        result
    }
}

/// Canonicalize 最近的现有祖先，再原样接回尚不存在的尾部。
fn canonicalize_with_missing_tail(path: &Path) -> Result<PathBuf, ToolError> {
    let mut ancestor = path.to_path_buf();
    let mut tail = Vec::new();
    while !ancestor.exists() {
        let name = ancestor
            .file_name()
            .ok_or_else(|| ToolError::new(format!("cannot resolve path: {}", path.display())))?;
        tail.push(name.to_os_string());
        if !ancestor.pop() {
            return Err(ToolError::new(format!(
                "cannot resolve path: {}",
                path.display()
            )));
        }
    }
    let mut resolved = ancestor
        .canonicalize()
        .map_err(|error| ToolError::new(format!("cannot resolve path: {error}")))?;
    for part in tail.into_iter().rev() {
        resolved.push(part);
    }
    Ok(resolved)
}
